### ISSUE TITLE
Linux: cannot open gdi files that contain special characters

### ISSUE THREAD

**Issue by danboid** [reicast/reicast-emulator#1006]
Linux reicast fails to open gdi files if any of the file names contained within contain special characters such as parentheses or commas.

**Comment by goigle** [comment]
Which Linux are you on?
I'm on Raspbian and my filenames have spaces and it works fine.

**Comment by danboid** [comment]
That's strange - I doubt the distro would matter but I run Arch.

**Comment by danboid** [comment]
Hi goigle

I have returned to try reicast again and I'm trying to run the same roms as last time. I noticed that the filenames of the files that make up the ROM contain not only spaces but commas and parentheses so it is more likely one of them that reicast is choking on. Hence I have edited the title of this issue to reflect that.

**Comment by Sarkie** [comment]
Give an example filename...

**Comment by danboid** [comment]
One such problematic gdi attached, albeit with an added .txt extension 

[Gauntlet Legends (Europe) (En,Fr,De).gdi.txt](https://github.com/reicast/reicast-emulator/files/1411018/Gauntlet.Legends.Europe.En.Fr.De.gdi.txt)

It reads:

5
1      0 4 2352 "Gauntlet Legends (Europe) (En,Fr,De) (Track 1).bin" 0
2    607 0 2352 "Gauntlet Legends (Europe) (En,Fr,De) (Track 2).bin" 0
3  45000 4 2352 "Gauntlet Legends (Europe) (En,Fr,De) (Track 3).bin" 0
4 245774 0 2352 "Gauntlet Legends (Europe) (En,Fr,De) (Track 4).bin" 0
5 246300 4 2352 "Gauntlet Legends (Europe) (En,Fr,De) (Track 5).bin" 0

**Comment by LoungeKatt** [comment]
Is this still an issue?

**Comment by danboid** [comment]
Yes, this is still an issue. I've just tried Sonic Adventure under the latest git version of reicast under Ubuntu 18.04. The gdi file as downloaded off archive.org looked like this, which didn't load:

```
3
1     0 4 2352 "Sonic Adventure (Europe) (En,Ja,Fr,De,Es) (Track 1).bin" 0
2 11361 0 2352 "Sonic Adventure (Europe) (En,Ja,Fr,De,Es) (Track 2).bin" 0
3 45000 4 2352 "Sonic Adventure (Europe) (En,Ja,Fr,De,Es) (Track 3).bin" 0
```

So I renamed the files and edited the .gdi file to look like this:

```
3
1     0 4 2352 SonicAdventureTrack1.bin 0
2 11361 0 2352 SonicAdventureTrack2.bin 0
3 45000 4 2352 SonicAdventureTrack3.bin 0
```

Then it loaded.

First I tried removing the brackets from the filenames, that didn't fix it. Next I tried removing the spaces with no luck. It wasn't until I removed the quotes that reicast sucessfully found and loaded al three bin files. With the quotes around  the .bin filenames it only seemed to load the first .bin file so I would only see the DC BIOS menu and the game failed to load.

**Comment by baka0815** [comment]
Looks like the issue could be line 112 in gdi.cpp:
https://github.com/reicast/reicast-emulator/blob/1f088f211d273b86875b91b567bd08a6a9fce0fe/core/imgread/gdi.cpp#L112

**Comment by LoungeKatt** [comment]
@baka0815 That wouldn't be the issue. That would still detect that the last character is a whitespace between both `somefile 0` and `"somefile" 0`
It's the next few lines that more likely need some tweaking.
```
if (last == '"')
		{
			for(;;) {
				gdi >> std::noskipws >> last;
				if (last == '"')
					break;
				track_filename += last;
			}
		}
		else
		{
			gdi >> track_filename;
			track_filename = last + track_filename;
		}

		gdi >> OFFSET;
```

**Comment by LoungeKatt** [comment]
I am curious of both of these have identical results in regards to this issue:

```
		if (last == '"')
		{
			for(;;) {
				gdi >> std::noskipws >> last;
				if (last == '"')
					break;
				track_filename += last;
			}
		}
```

```
		if (last == '"')
		{
			gdi >> std::noskipws >> last;
			for(;;) {
				if (last == '"')
					break;
				track_filename += last;
			}
			gdi >> std::noskipws >> last;
		}
```

**Comment by baka0815** [comment]
Maybe I'll have time tomorrow to look into this.

**Comment by baka0815** [comment]
It's all because the ``skipws``-flag is not reset...

**Comment by LoungeKatt** [comment]
Back when this was ported based on claims that RA could handle whitepsace and reicast couldn't, it was a simple oversight.
```
			gdi >> std::noskipws;
			for(;;) {
				gdi >> last;
				if (last == '"')
					break;
				track_filename += last;
			}
			gdi >> std::skipws;
```

So rather than a lot of changes to initialize variables and then explicitly define whitespace outside the actual usage, couldn't you just add the `gdi >> std::skipws;` after `track_filename += last;` and be done with it? Or even duplicate the original implementation by moving it around the `for`?

**Comment by baka0815** [comment]
Ok, I resubmitted the fix in #1191.

**Comment by LoungeKatt** [comment]
https://github.com/baka0815/reicast-emulator/blob/41ee5707a827ed94327b4d18551bcadc3f06c524/core/imgread/gdi.cpp#L117 should also be moved above https://github.com/baka0815/reicast-emulator/blob/41ee5707a827ed94327b4d18551bcadc3f06c524/core/imgread/gdi.cpp#L116 (for consistency)
